man/systemd.resource-control.xml

```diff
@@ -1426,6 +1426,76 @@ DeviceAllow=/dev/loop-control
 
         <xi:include href="version-info.xml" xpointer="v254"/></listitem>
       </varlistentry>
+
+      <varlistentry>
+        <term><varname>NFTSet=</varname><replaceable>family</replaceable>:<replaceable>table</replaceable>:<replaceable>set</replaceable></term>
+        <listitem>
+          <para>This setting provides a method for integrating dynamic cgroup IDs into firewall rules with
+          <ulink url="https://netfilter.org/projects/nftables/index.html">NFT</ulink> sets. The benefit of
+          using this setting is to be able to use the IDs as selectors in firewall rules easily and this in
+          turn allows more fine grained filtering. NFT rules for cgroup matching use numeric cgroup IDs,
+          which change every time a service is restarted, making them hard to use in systemd environment
+          otherwise.</para>
+
+          <para>This option expects a whitespace separated list of NFT set definitions. Each definition
+          consists of a colon-separated tuple of source type (only <literal>cgroup</literal>), NFT address
+          family (one of <literal>arp</literal>, <literal>bridge</literal>, <literal>inet</literal>,
+          <literal>ip</literal>, <literal>ip6</literal>, or <literal>netdev</literal>), table name and set
+          name. The names of tables and sets must conform to lexical restrictions of NFT table names. The
+          type of the element used in the NFT filter must match the type implied by the directive
+          (<literal>cgroup</literal>) as shown in the table below. When a control group is realized, the
+          corresponding ID will be appended to the NFT sets and it will be be removed when the control group
+          is removed. <command>systemd</command> only inserts elements to (or removes from) the sets, so the
```

Comment:
why does systemd have to remove the cgroup ids? why doesn#t nft clean them up on its own when the cgroup goes away?

Reply:
I don't think NFT subsystem in the kernel removes anything from any sets.

Reply:
so its cgroup sets will possibly carry all kinds of invalid ids?

Reply:
Why would they? PID1 will add IDs to the sets and remove them. If something else adds invalid IDs to the set, PID1 will not touch it.